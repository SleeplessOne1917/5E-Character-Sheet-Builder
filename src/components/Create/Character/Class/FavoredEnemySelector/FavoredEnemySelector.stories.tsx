import { ComponentMeta, ComponentStory } from '@storybook/react';

import FavoredEnemySelector from './FavoredEnemySelector';
import { Provider } from 'react-redux';
import { getTestStore } from '../../../../../redux/store';

export default {
	title: 'Components/CharacterCreation/Class/FavoredEnemySelector',
	component: FavoredEnemySelector,
	args: {
		values: ['ABERRATION'],
		monsters: {
			monsters: [
				'ABERRATION',
				'BEAST',
				'CELESTIAL',
				'CONSTRUCT',
				'DRAGON',
				'ELEMENTAL',
				'FEY',
				'FIEND',
				'GIANT',
				'HUMANOID',
				'MONSTROSITY',
				'OOZE',
				'PLANT',
				'UNDEAD'
			],
			humanoids: [
				'DWARF',
				'ELF',
				'GNOLL',
				'GNOME',
				'GOBLINOID',
				'HUMAN',
				'KOBOLD',
				'LIZARDFOLK',
				'MERFOLK',
				'ORC',
				'SAHUAGIN'
			]
		}
	},
	decorators: [
		Story => (
			<Provider store={getTestStore()}>
				<Story />
			</Provider>
		)
	]
} as ComponentMeta<typeof FavoredEnemySelector>;

const Template: ComponentStory<typeof FavoredEnemySelector> = args => (
	<FavoredEnemySelector {...args} />
);

export const Default = Template.bind({});

export const Humanoid = Template.bind({});

Humanoid.args = {
	values: ['DWARF', 'ELF']
};
