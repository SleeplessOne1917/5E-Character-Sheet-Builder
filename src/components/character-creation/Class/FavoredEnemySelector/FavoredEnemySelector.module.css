.container {
	border: 3px solid var(--davy);
	background-color: var(--heaven-light);
	border-radius: 1rem;
	display: flex;
	flex-direction: column;
	align-items: center;
	padding: 0.5rem;
	width: fit-content;
	margin: 1rem;
}

.humanoid-container {
	display: flex;
}

.label {
	font-size: 1.3rem;
	padding: 0.3rem;
}

.monster-type-container {
	display: flex;
	flex-direction: column;
	align-items: center;
	padding: 0.5rem;
}