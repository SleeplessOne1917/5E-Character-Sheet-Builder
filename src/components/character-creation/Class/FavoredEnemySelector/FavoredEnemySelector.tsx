import { MonsterSubtype, MonsterType } from '../../../../types/srd';

import Select from '../../../Select/Select';
import classes from './FavoredEnemySelector.module.css';
import { useAppSelector } from '../../../../hooks/reduxHooks';
import { useCallback } from 'react';

type FavoredEnemySelectorProps = {
	values: (MonsterType | MonsterSubtype | null)[];
	onChange: (values: (MonsterType | MonsterSubtype | null)[]) => void;
	monsters: { humanoids: MonsterSubtype[]; monsters: MonsterType[] };
};

const FavoredEnemySelector = ({
	values,
	onChange,
	monsters
}: FavoredEnemySelectorProps) => {
	const handleChange = useCallback(
		(value: MonsterType) => {
			if (value === 'HUMANOID' && values.length === 1) {
				onChange([null, null]);
			} else if (value !== 'HUMANOID') {
				onChange([value]);
			}
		},
		[onChange, values.length]
	);

	const selectedTypes = useAppSelector(
		state => state.editingCharacter.classInfo.favoredEnemies
	)?.flatMap(val => val);

	return (
		<div
			className={`${classes.container}${
				!values.includes(null) ? ` ${classes.selected}` : ''
			}`}
			data-testid="favored-enemy-selector"
		>
			<div className={classes['monster-type-container']}>
				<label id="monster-type" className={classes.label}>
					Select Monster Type
				</label>
				<Select
					labelledBy="monster-type"
					onChange={value =>
						handleChange((value === 'blank' ? null : value) as MonsterType)
					}
					options={[{ label: '\u2014', value: 'blank' }].concat(
						monsters.monsters
							.filter(m => !selectedTypes?.includes(m) || values.includes(m))
							.map(m => ({ label: m[0] + m.slice(1).toLowerCase(), value: m }))
					)}
					value={
						values.length > 1
							? 'HUMANOID'
							: values[0] === null
							? 'blank'
							: values[0]
					}
				/>
			</div>
			{values.length > 1 && (
				<div className={classes['humanoid-container']}>
					<div className={classes['monster-type-container']}>
						<label className={classes.label} id="humanoid-1">
							Select Humanoid Type 1
						</label>
						<Select
							labelledBy="humanoid-1"
							onChange={value =>
								onChange([
									(value === 'blank' ? null : value) as MonsterSubtype,
									values[1]
								])
							}
							options={[{ value: 'blank', label: '\u2014' }].concat(
								monsters.humanoids
									.filter(h => !selectedTypes?.includes(h) || values[0] === h)
									.map(h => ({
										value: h,
										label: h[0] + h.slice(1).toLowerCase()
									}))
							)}
							value={values[0] ?? 'blank'}
						/>
					</div>
					<div className={classes['monster-type-container']}>
						<label className={classes.label} id="humanoid-2">
							Select Humanoid Type 2
						</label>
						<Select
							labelledBy="humanoid-2"
							onChange={value =>
								onChange([
									values[0],
									(value === 'blank' ? null : value) as MonsterSubtype
								])
							}
							options={[{ value: 'blank', label: '\u2014' }].concat(
								monsters.humanoids
									.filter(h => !selectedTypes?.includes(h) || values[1] === h)
									.map(h => ({
										value: h,
										label: h[0] + h.slice(1).toLowerCase()
									}))
							)}
							value={values[1] ?? 'blank'}
						/>
					</div>
				</div>
			)}
		</div>
	);
};

export default FavoredEnemySelector;
