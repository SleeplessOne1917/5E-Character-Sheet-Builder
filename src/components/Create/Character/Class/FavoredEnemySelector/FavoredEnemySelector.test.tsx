import { render, screen } from '@testing-library/react';
import * as stories from './FavoredEnemySelector.stories';
import { composeStories } from '@storybook/testing-react';

const { Default, Humanoid } = composeStories(stories);

describe('renders correctly', () => {
	it('when default', () => {
		render(<Default />);

		expect(screen.getByTestId('favored-enemy-selector')).toMatchSnapshot();
	});

	it('when humanoid', () => {
		render(<Humanoid />);

		expect(screen.getByTestId('favored-enemy-selector')).toMatchSnapshot();
	});
});
