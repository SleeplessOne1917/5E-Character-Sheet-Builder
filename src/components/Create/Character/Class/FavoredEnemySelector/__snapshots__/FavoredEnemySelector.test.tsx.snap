// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when default 1`] = `
<div
  class="container selected"
  data-testid="favored-enemy-selector"
>
  <div
    class="monster-type-container"
  >
    <div
      class="list-box-and-error-container"
      data-testid="select"
    >
      <label
        id="monster-type-label"
        style="font-size: 1.3rem;"
      >
        Select Monster Type
      </label>
      <div
        aria-labelledby="monster-type-label"
        class="list-box-container"
        style="font-size: 1.2rem;"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          class="button"
          data-testid="select-button"
          type="button"
        >
          Aberration
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <ul
          aria-activedescendant="Aberration"
          class="list"
          data-testid="select-list"
          role="listbox"
          style="display: none;"
          tabindex="-1"
        >
          <li
            aria-selected="false"
            class="item"
            data-testid="data-blank"
            role="option"
            tabindex="0"
          >
            —
          </li>
          <li
            aria-selected="true"
            class="item selected"
            data-testid="data-ABERRATION"
            role="option"
            tabindex="0"
          >
            Aberration
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-BEAST"
            role="option"
            tabindex="0"
          >
            Beast
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-CELESTIAL"
            role="option"
            tabindex="0"
          >
            Celestial
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-CONSTRUCT"
            role="option"
            tabindex="0"
          >
            Construct
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-DRAGON"
            role="option"
            tabindex="0"
          >
            Dragon
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-ELEMENTAL"
            role="option"
            tabindex="0"
          >
            Elemental
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-FEY"
            role="option"
            tabindex="0"
          >
            Fey
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-FIEND"
            role="option"
            tabindex="0"
          >
            Fiend
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-GIANT"
            role="option"
            tabindex="0"
          >
            Giant
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-HUMANOID"
            role="option"
            tabindex="0"
          >
            Humanoid
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-MONSTROSITY"
            role="option"
            tabindex="0"
          >
            Monstrosity
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-OOZE"
            role="option"
            tabindex="0"
          >
            Ooze
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-PLANT"
            role="option"
            tabindex="0"
          >
            Plant
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-UNDEAD"
            role="option"
            tabindex="0"
          >
            Undead
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly when humanoid 1`] = `
<div
  class="container selected"
  data-testid="favored-enemy-selector"
>
  <div
    class="monster-type-container"
  >
    <div
      class="list-box-and-error-container"
      data-testid="select"
    >
      <label
        id="monster-type-label"
        style="font-size: 1.3rem;"
      >
        Select Monster Type
      </label>
      <div
        aria-labelledby="monster-type-label"
        class="list-box-container"
        style="font-size: 1.2rem;"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          class="button"
          data-testid="select-button"
          type="button"
        >
          Humanoid
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <ul
          aria-activedescendant="Humanoid"
          class="list"
          data-testid="select-list"
          role="listbox"
          style="display: none;"
          tabindex="-1"
        >
          <li
            aria-selected="false"
            class="item"
            data-testid="data-blank"
            role="option"
            tabindex="0"
          >
            —
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-ABERRATION"
            role="option"
            tabindex="0"
          >
            Aberration
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-BEAST"
            role="option"
            tabindex="0"
          >
            Beast
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-CELESTIAL"
            role="option"
            tabindex="0"
          >
            Celestial
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-CONSTRUCT"
            role="option"
            tabindex="0"
          >
            Construct
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-DRAGON"
            role="option"
            tabindex="0"
          >
            Dragon
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-ELEMENTAL"
            role="option"
            tabindex="0"
          >
            Elemental
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-FEY"
            role="option"
            tabindex="0"
          >
            Fey
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-FIEND"
            role="option"
            tabindex="0"
          >
            Fiend
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-GIANT"
            role="option"
            tabindex="0"
          >
            Giant
          </li>
          <li
            aria-selected="true"
            class="item selected"
            data-testid="data-HUMANOID"
            role="option"
            tabindex="0"
          >
            Humanoid
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-MONSTROSITY"
            role="option"
            tabindex="0"
          >
            Monstrosity
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-OOZE"
            role="option"
            tabindex="0"
          >
            Ooze
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-PLANT"
            role="option"
            tabindex="0"
          >
            Plant
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="data-UNDEAD"
            role="option"
            tabindex="0"
          >
            Undead
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="humanoid-container"
  >
    <div
      class="monster-type-container"
    >
      <div
        class="list-box-and-error-container"
        data-testid="select"
      >
        <label
          id="humanoid-1-label"
          style="font-size: 1.5rem;"
        >
          Select Humanoid Type 1
        </label>
        <div
          aria-labelledby="humanoid-1-label"
          class="list-box-container"
          style="font-size: 1.3rem;"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="button"
            data-testid="select-button"
            type="button"
          >
            Dwarf
            <svg
              aria-hidden="true"
              class="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <ul
            aria-activedescendant="Dwarf"
            class="list"
            data-testid="select-list"
            role="listbox"
            style="display: none;"
            tabindex="-1"
          >
            <li
              aria-selected="false"
              class="item"
              data-testid="data-blank"
              role="option"
              tabindex="0"
            >
              —
            </li>
            <li
              aria-selected="true"
              class="item selected"
              data-testid="data-DWARF"
              role="option"
              tabindex="0"
            >
              Dwarf
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-ELF"
              role="option"
              tabindex="0"
            >
              Elf
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-GNOLL"
              role="option"
              tabindex="0"
            >
              Gnoll
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-GNOME"
              role="option"
              tabindex="0"
            >
              Gnome
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-GOBLINOID"
              role="option"
              tabindex="0"
            >
              Goblinoid
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-HUMAN"
              role="option"
              tabindex="0"
            >
              Human
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-KOBOLD"
              role="option"
              tabindex="0"
            >
              Kobold
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-LIZARDFOLK"
              role="option"
              tabindex="0"
            >
              Lizardfolk
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-MERFOLK"
              role="option"
              tabindex="0"
            >
              Merfolk
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-ORC"
              role="option"
              tabindex="0"
            >
              Orc
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-SAHUAGIN"
              role="option"
              tabindex="0"
            >
              Sahuagin
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="monster-type-container"
    >
      <div
        class="list-box-and-error-container"
        data-testid="select"
      >
        <label
          id="humanoid-2-label"
          style="font-size: 1.3rem;"
        >
          Select Humanoid Type 2
        </label>
        <div
          aria-labelledby="humanoid-2-label"
          class="list-box-container"
          style="font-size: 1.2rem;"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="button"
            data-testid="select-button"
            type="button"
          >
            Elf
            <svg
              aria-hidden="true"
              class="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <ul
            aria-activedescendant="Elf"
            class="list"
            data-testid="select-list"
            role="listbox"
            style="display: none;"
            tabindex="-1"
          >
            <li
              aria-selected="false"
              class="item"
              data-testid="data-blank"
              role="option"
              tabindex="0"
            >
              —
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-DWARF"
              role="option"
              tabindex="0"
            >
              Dwarf
            </li>
            <li
              aria-selected="true"
              class="item selected"
              data-testid="data-ELF"
              role="option"
              tabindex="0"
            >
              Elf
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-GNOLL"
              role="option"
              tabindex="0"
            >
              Gnoll
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-GNOME"
              role="option"
              tabindex="0"
            >
              Gnome
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-GOBLINOID"
              role="option"
              tabindex="0"
            >
              Goblinoid
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-HUMAN"
              role="option"
              tabindex="0"
            >
              Human
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-KOBOLD"
              role="option"
              tabindex="0"
            >
              Kobold
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-LIZARDFOLK"
              role="option"
              tabindex="0"
            >
              Lizardfolk
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-MERFOLK"
              role="option"
              tabindex="0"
            >
              Merfolk
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-ORC"
              role="option"
              tabindex="0"
            >
              Orc
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-SAHUAGIN"
              role="option"
              tabindex="0"
            >
              Sahuagin
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
